Use Ember shims for redux packages instead of ember-browserify

sandersky opened the pull request:
In an effort to remove the dependency of `ember-browserify` I've created three new shims for packages previously being included via `ember-browserify`:

* [ember-redux-shim](https://github.com/ciena-blueplanet/ember-redux-shim)
* [ember-redux-thunk](https://github.com/ciena-blueplanet/ember-redux-thunk)
* [ember-redux-saga](https://github.com/ciena-blueplanet/ember-redux-saga)

This pull request is an alternative approach to what has been done in #57

sandersky commented:
If you like this solution I'm happy to release 1.x versions of the shims so they aren't hard-pinned to a point release.

toranb commented:
@sandersky wow you really went all out today :)

I'll consider this as one of the viable options to help us migrate away from browserify. Sorry for the delay in reply

MiguelMadero commented:
@sandersky this looks great, but I feel like I'm missing some magic. On https://github.com/toranb/ember-redux/pull/57 we had to use `transformation: amd`, but on the shim, it seems like we only merge the tree (https://github.com/ciena-blueplanet/ember-redux-thunk/blob/master/index.js#L14). Is there an additional step that we're missing? Is there another addon that makes the shim work?

MiguelMadero commented:
Oh, those are just ES modules, so we're transforming them already along with other addon code. So the main difference is that treeForAddon will transpile, while treeForVendor wasn't, so the transformation from AMD was needed for vendors, but for addon code is taken care of by babel. 

This looks really good and I like that we don't have to depend on Ember CLI 2.9

toranb commented:
@MiguelMadero good timing - my feeling is that if shims help us avoid the `ember-cli v2.9` requirement I'd prefer that approach. The only downside that I can think of off the top is ownership of the shims - @sandersky you open to adding Miguel and I to the shim projects if we adopt this approach (to increase the bus factor to 3 instead of 1) ?

MiguelMadero commented:
When you asked about concerns with both approaches   bus factor was the main one I could think of for shims. 

The shims seem well done and straight forward, @sandersky did a great job and I like the CI setup and deployment. If we can increase the bus factor Im :thumbsups: for this approach. 

Im afk, but would like to look at the compiled payload. Im not expecting surprises, but still would like to look at it. 

Another smaller advantage here is that once/if we have tree-shaking in cli, it would just work here. If we imported and amd "prebuilt" package we would have to switch to src for tree shaking to work.

dustinfarris commented:
I like bringing in the dependency code directly, just have to be careful redux/-thunk/-saga don't add some weird babel plugin requirement in the future.  Ultimately, with this strategy, it is ember-redux user apps that will be transpiling everything, so if they are not set up with everything redux needs to compile, they will start seeing weird errors that may not make sense.  Luckily, the reactjs organization seems pretty sensitive to this sort of thing in general.

@toranb I think the shims could be done as in-repo-addons to keep it under the same project.  They would live in lib/ instead of as separate npm dependencies.  @sandersky is there any reason that wouldn't work?

MiguelMadero commented:
The in-repo idea is interesting and I can't think of a reason to use ember-redux-shim without ember-redux. As a side note, the name can cause a bit of confusion, isn't really a shim for `ember-redux`, it provides a shim to make redux work inside of ember. One problem I see for in-repo is using other shims or using redux-saga, which is right now only used for tests. 

One thing I like about the shims is that it provides a relatively easy pattern to follow for other addons, for example, we use redux-logger, graphql-tag, redux-actions and the apollo-client for GraphQL, which also depends on browserify today, so when we un-browserify our app, we will create more shims.

MiguelMadero commented:
@sandersky today, `ember-redux` only imports redux-saga for tests (see [transofmation](https://github.com/toranb/ember-redux/pull/57/files#diff-168726dbe96b3ce427e7fedce31bb0bcR20), [rollup](https://github.com/toranb/ember-redux/pull/68/files#diff-168726dbe96b3ce427e7fedce31bb0bcR17) and [master](https://github.com/toranb/ember-redux/blob/8dba402c5e8eae010af35d3996feb25689da1bd9/tests/dummy/app/sagas/counter.js#L2)

I'm not sure what's a good solution for this using shims. We could include the addon only if we're in tests by using [shouldIncludeChildAddon](https://ember-cli.com/api/classes/Addon.html#method_shouldIncludeChildAddon). I'm not sure if this will *always* remove the addon or if it removes it only if the addon isn't explicitly added by the app. For example, an app may want to use `ember-redux-saga` directly even though `ember-redux` says that we shouldn't include it. Would that work.

MiguelMadero commented:
> Im afk, but would like to look at the compiled payload. Im not expecting surprises, but still would like to look at it.

@toranb @sandersky I looked at the code. It all looks fine, just like what you would expect from babel/CLI, even a bit better than webpack's output. There is one small thing that I want to raise, I don't think it's an issue, but I don't fully understand. redux includes the following code:

```
if ('production' !== 'production' && typeof isCrushed.name === 'string' && isCrushed.name !== 'isCrushed') {
    (0, _reduxUtilsWarning['default'])('You are currently using minified code outside of NODE_ENV === \'production\'. ' + 'This means that you are running a slower development build of Redux. ' + 'You can use loose-envify (https://github.com/zertosh/loose-envify) for browserify ' + 'or DefinePlugin for webpack (http://stackoverflow.com/questions/30030031) ' + 'to ensure you have the correct code for your production build.');
  }
```

On production builds removes all that because it will never hit it (since `'production' !== 'production'` will short-circuit it). I'm not super familiar with webpack, but the comment implies that there may be other optimizations that webpack does that we might not. Maybe all of that is babel and we're fine, but I'm not familiar enough with this. @sandersky is there something important we may be missing? I noticed you [explicitly replaced the string](https://github.com/ciena-blueplanet/ember-redux-shim/blob/master/index.js#L20) so it uses `production`

sandersky commented:
Currently I'm included `ember-redux-saga` as a `devDependency` and not adding it to the blueprints so consumers won't get it. If they want `ember-redux-saga` then they can ember install it into their app/addon. Regarding the production !== production check I'm not sure of a good way to remove that code during a babel build but would love to do so if anyone has ideas. Also happy to add more people as maintainers of the shims.

sandersky commented:
I can also improve the hard-coded production check to use the consumers environment as well like I've done here https://github.com/ciena-blueplanet/ember-seamless-immutable-shim/blob/master/index.js#L40

sandersky commented:
Assuming older versions of redux are in CommonJS format my approach can handle that as well as done here: https://github.com/ciena-blueplanet/ember-validator-shim/blob/master/index.js

sandersky commented:
@MiguelMadero and @toranb I've given both of you write access to the 3 shims.

MiguelMadero commented:
Thanks @sandersky and thanks for clarifying. 
BTW, the rewrite is fine, when we minify the code the actual function will be removed. 

re: older versions I'll defer to @toranb and others about this. I'm not sure if this is an actually issue, but it's good to know there are options.

toranb commented:
@MiguelMadero @sandersky I don't think we will likely support anything < redux v3.5 or v3.6

If that is the case, what all is blocking us from taking the shim approach forward? And what considerations do we need to make regarding semver? Will this be a v2.0 breaking change or ?

dustinfarris commented:
Only real change I can think of: anyone importing redux/-thunk via browserify on their own needs to know they now have 2 copies of redux/-thunk in their vendor bundle until they remove instances of "npm:" from their code, and ember-browserify from package.json if not needed for anything else.

Just something to mention in the release notes.

Also needs rebase.

MiguelMadero commented:
I don't see any blockers. It works on my machine and I tried that in one our project. 

re: v2.0/breaking change. It's debatable, but IMO, it seems like an interface change. We were importing `npm:redux` explicitly and I had to update all the imports, not only to reduce the payload size, but something wasn't working. I didn't look too deep into the issue, maybe just PEBCAK, but I can have another look tomorrow if that would make a difference.

dustinfarris commented:
Probably stemming from the two copies of redux in your vendor bundle before you remove "npm:" prefixes.  So potentially two redux stores (one probably not initialized). when you npmRedux.dispatch and redux.dispatch you get different results, etc..

All the more reason to make this clear in the release notes; and since this actually breaks, sounds like a major bump after all.

> On Dec 16, 2016, at 2:34 PM, Miguel Madero <notifications@github.com> wrote:
> 
> I don't see any blockers. It works on my machine and I tried that in one our project.
> 
> re: v2.0/breaking change. It's debatable, but IMO, it seems like an interface change. We were importing npm:redux explicitly and I had to update all the imports, not only to reduce the payload size, but something wasn't working. I didn't look too deep into the issue, maybe just PEBCAK, but I can have another look tomorrow if that would make a difference.
> 
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub <https://github.com/toranb/ember-redux/pull/63#issuecomment-267527499>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ABCWvcj3RZBYwtUrt_LBB-wzGat6AKMJks5rIjDogaJpZM4LLHnU>.
> 

-- 
Dustin Farris

MiguelMadero commented:
@dustinfarris yes, that was likely the problem.

toranb commented:
@dustinfarris @MiguelMadero @sandersky as I'm reading this over and it's clear we will do v2 bump when we merge this in the coming week or so. 

What did everyone decide on the "in-app repo" for shims (redux/redux-thunk) vs external? One pro is that we suddenly have what looks like a unified version (as they are all in this repo/in sync/ change together). One con of not doing this can be version/dependency hell

If we included them in this repo could they still be published (and thus consumed) independently as if they were standalone? I ask because this option would provide the best of both worlds in some regard (publishing from here as versions change /others could then consume those or use ember-twiddle with them independent of the `ember-redux` addon.

@sandersky I see your mention of `ember-frost-bunsen` and assume that is show stopper here (likely because if you use both `ember-frost-bunsen` and `ember-redux` you would get duplicate redux code?) *correct me if I'm wrong here

The concerns around `ember-redux-saga` are a non issue now that I see it's not included in the blueprint (and thus not installed for people when using `ember install ember-redux`)

If we decide to keep the shims external I think it's still very much the right direction for v2 - I just wanted to understand and maybe explore this further as I don't take a major version bump lightly. I want to see what this looks like from both sides of the coin/ and also consider what this would mean for ember-twiddle (as one of our main strengths to this project is that people can drop in and hack redux/ember in a twiddle quick)

Any considerations when using ember-twiddle that we haven't talked about yet? I assume you just add it like you would another other addon and use `redux` not `npm:redux`

toranb commented:
@sandersky did you alter the redux shim to work around the hard-coded production check? also -is that something we should consider for thunk/or saga also?

dustinfarris commented:
I think the shim is a good idea in general for people who want to play with redux (or any other non-AMD) in ember, but are not interested in ember-redux.  I'm glad they exist.

@toranb What do you mean by "unified version"?  in-repo-addon or not, the user must install redux separately.  We automate this initially with a default blueprint (as does ember-redux-shim); but once everything is installed, it is the user's responsibility to keep redux up-to-date.  Upgrading ember-redux or ember-redux-shim does not upgrade redux—an in-repo-addon would not change that.

----

Really, the main difference is the impact on the user's package.json.  Without using in-repo-addons, when the user does 

```
ember install ember-redux
```

then their package.json diff is something like:

```diff
@@ package.json
+"ember-redux": "^2.0.0",
+"ember-redux-shim": "1.0.0",
+"ember-redux-thunk": "1.0.0",
+"redux": "3.6.0",
+"redux-thunk": "2.1.0",
```

So the user now has 5 dependencies to maintain instead of 3.

----

- ember-twiddle demo would need to be updated, but would not be compromised by either approach
- you could technically npm publish the shims from within the in-repo-addon, although i'm not sure taking on that overhead is a net win for ember-redux

MiguelMadero commented:
While I like the idea of a simpler installation and seeing fewer dependencies in package.json, I think it still makes more sense to keep the shims external to avoid code duplication, having two instances of the shim would lead to including redux more than once in consumer code if someone had both `emerber-redux` with an embedded shim and `ember-redux-shim`. I think it's better to keep them 1:1 and make the shim the only way to import a lib. Another slight problem may be if add-ons have it as a dependency (instead of devDependency and relying on a blueprint) since the treeForAddon would execute N times, each time leading to more code in vendor.js (that is until we have https://github.com/ember-cli/ember-cli/pull/6573) 

All that say, this may sound like more trouble than what is worth, but if we wanted to avoid code duplication and extra entries in package.json, I'm wondering if a possible solution is to make the call to treeForAddon on each shim idempotent and making it the shims `dependency` of this addon instead of relying on the blueprints to make it a dependency on the consumer app/addon. That way the redux code would end up only once in vendor.js regardless of other add-ons that may have it as a dependency and also the consumer wouldn't have to worry about adding it explicitly, but even if they did, that wouldn't result in duplication. As I said, it might not be worth it. 

@toranb did you consider releasing as 2.0-beta ? That might make it easier to iterate on this and get some additional feedback.

sandersky commented:
@toranb I've cut a PR on `ember-redux-shim` to take into account the consumers environment. https://github.com/ciena-blueplanet/ember-redux-shim/pull/2 I'll probably get back to this tomorrow to button up any remaining items.

sandersky commented:
Regarding lodash 4, it is a redux dependency https://github.com/reactjs/redux/blob/master/package.json#L68-L69.

Regarding npm access, the CI automatically publishes to npm on merged PR's for the shims. My lodash shim also exports files under the 3.x paths and works pretty well with consumers of both versions.

Regarding duplicates I'm not sure what is duplicated if addons just specify things in blueprints and devDependencies, giving consumers control of exact versions.

MiguelMadero commented:
> if addons just specify things in blueprints and devDependencies

Yes, that's the `if` I'm questioning. This is a mute point if we can assume that's the expected behavior of an addon and using dependencies was considered a bug on the addon and not a problem on the shims. However, I have seen many addons rely on dependencies in this past, this [RFC to cache addontTree*](https://github.com/ember-cli/rfcs/blob/addon-tree-caching/active/0000-addon-tree-caching.md#motivation) got me thinking more about this problem. In their example, ember-wormhole is a dependency of ember-basic-dropdown (and others), so it gets built multiple times, leading to slower builds. 

*UPDATE:* I just had another look at the shim and looks pretty similar to the wormhole scenario, its just more code from the tree and [other code from the addon would overwrite redux](https://github.com/ciena-blueplanet/ember-redux-shim/blob/master/index.js#L50). So my new theory is that, if anyone uses this as a dependency, that only means that we build the shim and ember-redux code twice, which is the same issue that the RFC ^^ and https://github.com/ember-cli/ember-cli/pull/6573 address. Since this is only "more code" for the addon, it should behave the same as "more of the same versions of wormhole". I'll double check all of this tomorrow, but I don't think this is a blocker to release beta, this is just to satisfy my own curiosity.

sandersky commented:
@MiguelMadero if I'm understanding correctly what you are talking about is the redux code could potentially be built multiple times when a consumer builds their app if multiple addons are including the redux shim. However in the end only one copy of the redux code would land in their build because the redux shim builds would overwrite one another. So in the end an apps build would only contain one copy but the process of building could be potentially slower as it would be doing the work multiple times. Does that sounds correct?

The UI team at Ciena, where I work, uses the blueprint/devDepdency approach on all of our repos under [ciena-frost](https://github.com/ciena-frost/) as well as [ciena-blueplanet](https://github.com/ciena-blueplanet/) and it has proven to be a good approach so far. The reason we chose this route is we try to use a very modular design with our addons; addons consuming addons consuming addons; and many addons depend on some of the same addons. Before going the blueprint route we'd end up with X on version 1 of Y, Z on version 2 of Y, etc, which proved to be dependency hell when upgrading things. The blueprint approach eased this pain as it allows our consuming apps to pin exactly which versions of dependencies we want. Of course this could potentially be problematic as well if a consumer starts blindly upgrading addons to versions that are incompatible with other addons they use. But we've muted that argument by deeming testing important. If you have proper tests in place you should be able to blindly upgrade things w/o any major functional side effects, as your tests would catch those.

sandersky commented:
Also I'm all for having an `ember-redux` Github organization as it would be nice to house all of this under one roof and make managing access rights much easier.

dbashford commented:
+1 on the ember-redux org, made even more useful with the introduction of the shim repos.

MiguelMadero commented:
> if I'm understanding correctly what you are talking about is the redux code could potentially be built multiple times 

I originally thought the issue could lead to duplicated code in vendor.js. Now I'm thinking that it would only result in slower builds, which isn't really our problem to solve. I would like to confirm this, but it's certainly not a blocker. 

We're also using blueprint/devDependencies at Zenefits for the same reason, I was a bit uncertain if that was the best way to go, but it's working fine so far. I have seen some add-ons going the other route. 

+1 for ember-redux org

dustinfarris commented:
@MiguelMadero aside from version mis-matches that @sandersky mentioned, I think the only time there is a real issue is with ember-browserify which prepends "npm:", so in vendor.js, so you could potentially have

```js
define('redux', ... );
define('npm:redux', ... );
```

which I believe is why you had problems earlier.

toranb commented:
@sandersky seems like all is good here after all. Could we grab one additional npm shim name? I have [ember-redux-saga-shim](https://www.npmjs.com/package/ember-redux-saga-shim) and I'm curious if we should grab ember-redux-thunk-shim to keep them in sync (all having the shim name). Thoughts?

dustinfarris commented:
IMHO I would like to see -shim added to all to convey purpose.  The way I read it, ember-<package>-shim gives you ability to import in your project whereas ember-<package> alone is more about additional functionality/integration/abstractions.


> On Dec 20, 2016, at 10:40 AM, Toran Billups <notifications@github.com> wrote:
> 
> @sandersky <https://github.com/sandersky> seems like all is good here after all. Could we grab one additional npm shim name? I have ember-redux-saga-shim <https://www.npmjs.com/package/ember-redux-saga-shim> and I'm curious if we should grab ember-redux-thunk-shim to keep them in sync (all having the shim name). Thoughts?
> 
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub <https://github.com/toranb/ember-redux/pull/63#issuecomment-268138869>, or mute the thread <https://github.com/notifications/unsubscribe-auth/ABCWveUonGdGc1cAJY1VTJHFhJL3KbTdks5rJ0AagaJpZM4LLHnU>.
> 

-- 
Dustin Farris

toranb commented:
@sandersky I think the next step(s) here are adding repos to the organization correct? this will allow us some time to publish betas and experiment with the correct build step before releasing v2.0

toranb commented:
@sandersky can we sync up this week (assuming you had some time). I'd like to start the v2 (beta) branch with a new PR that is focused on adding shims to replace browserify (as that was the consensus from what I see above^^). We can link to this for history around the decision but I'd like a new PR so we can work through implementation of this (as that is the next step). I see you have the repos added to the organization but I'd like to move forward planning a Feb 1 v2 release (or there abouts). 

Let me know on slack/or via email if you will have some bandwidth in the coming week or 2

toranb commented:
@sandersky thanks for the work this past week on all the shims/renames/and everything in between! A few more questions to see if we can make some progress in the coming week

1) can you rebase this after all the shim work from last week?
2) what version number scheme do you have planned for the shims after we nail this? match the version of the library we shim /or do our own version numbering?
3) do you need anything else from me to get the shims up and running with either the ember-redux organization/or the ember-redux npm user?

for those keeping up - we did end up creating the `ember-redux` github organization/ we have a github user (ember-redux-admin) / we have an npm user (ember-redux) / we have an email for the domain to setup accounts admin [at] ember-redux [dot] com

sandersky commented:
1. @toranb I've rebased with upstream and updated the PR to use the latest shims.
2. Regarding the version scheme for shims I would plan to release a `1.0.0` of each shim once we all agree everything is good to go, then as the npm package the shim is shimming updates, we'd apply updates to the shim in the same fashion. Basically just make sure we are following semver along with the npm packages being shimmed. For the case of `redux-saga` they aren't on a `1.0.0` release yet so we are probably better off not releasing a 1.x version of the shim yet. Currently the shims use blueprints to install the packages being shimmed but I'm open to changing those to npm dependencies if others prefer (for the case of shims I don't have a strong argument for one way over the other).
3. I think all that remains is everyone agreeing on how everything is done and if we want to use blueprints vs npm dependencies in the shims. Once that is all 100% in agreement I can make any necessary adjustments.

dustinfarris commented:
My vote is for blueprints.  I'm generally against burying core dependencies (as in redux vs. something like broccoli-merge-trees) if it can be avoided.  Keeps the dep tree flatter and explicitly puts the user in control of their redux (and company) versions.

Personally, I think this is an under-celebrated feature of ember-cli.

toranb commented:
@sandersky @MiguelMadero @dbashford I just published v2.0.0-beta.1

I just tried this on ember twiddle `"ember-redux": "2.0.0-beta.1"` and it seems to blow up w/ this

`No addon.js created`

I'm not sure if this is shim/or beta release (npm) related but I'll need to take a deeper look later in the week. I also found doing `ember install ember-redux@2.0.0-beta.1` blew up locally 

![screen shot 2017-01-10 at 8 44 04 pm](https://cloud.githubusercontent.com/assets/147411/21833142/b93ed430-d775-11e6-9923-34b3695d8466.png)

I'll dive in when I have a little more time this week unless someone beats me 2 it

toranb commented:
@sandersky was able to fix the issue above ^^ by doing this explicitly in the package.json (after the ember install)

"redux-thunk": "^2.1.0"

that required/ or would the shim install take care of it?
